llvm/lib/Transforms/Utils/LoopIterCountAssumptions.cpp

```diff
@@ -0,0 +1,245 @@
+                       "SCEVAddRecExpr.\nEncountered no ADDRec "
+                       "expressions in the Loop Compare Instruction.");
+  return std::nullopt;
+}
+
+/// Return the Branch Compare Instruction of CurrentLoop if the Loop is well
+/// formed and this pass can process the Predicate
+ICmpInst *getLoopCmpInst(Loop &CurrentLoop) {
+
+  /// Check that the loop has a single Exiting Block. If the CurrentLoop
+  /// contains multiple Exiting Block, ExitBB will be a nullptr
+  auto *ExitBB = CurrentLoop.getExitingBlock();
+  if (!ExitBB)
+    return nullptr;
+
+  BranchInst *BI = dyn_cast<BranchInst>(ExitBB->getTerminator());
+  if (!BI)
+    return nullptr;
+
+  ICmpInst *LoopCmpInstr = dyn_cast<ICmpInst>(BI->getCondition());
+  if (!LoopCmpInstr)
+    return nullptr;
+
+  LLVM_DEBUG(dbgs() << "Condition Found: " << *LoopCmpInstr << "\n");
+
+  // Do not process equal and not-equal Compare Instructions, since they do not
+  // work well with the currently used assumption-construction-method of reusing
+  // the ICmpInst Predicate
+  if (LoopCmpInstr->getPredicate() == CmpInst::ICMP_EQ ||
+      LoopCmpInstr->getPredicate() == CmpInst::ICMP_NE) {
+    LLVM_DEBUG(dbgs() << "LoopIterCountAssumptions-Warning: EQ and NE "
+                         "conditions are currently not supported!");
+    return nullptr;
+  }
+  return LoopCmpInstr;
+}
+
+// Insert a Minimum Iteration Count Assumption into the Loop Header
+void insertMinIterAssumption(ICmpInst &LoopCmpInstr, BasicBlock &LoopHeader,
+                             ScalarEvolution &SE,
+                             const BundledSCEV &BunldedSCEV, int MinIterCount,
+                             AssumptionCache &AC) {
+  Instruction *HeaderTerminator = LoopHeader.getTerminator();
+
+  // LoopRotate uses SimplifyQuery to determine, if a Branch is conditional or
+  // not. SimplifyQuery can only take an Assumption into account, if it is
+  // before the to-be-evaluated Compare Instruction. Thus assumptions have to be
+  // inserted before the LoopCmpInstr.
+  IRBuilder<> Builder(dyn_cast<Instruction>(&LoopCmpInstr));
+
+  SCEVExpander Expander(SE, LoopHeader.getModule()->getDataLayout(),
+                        "expanded");
+
+  // Get Minimum Iteration Value Expression
+  const SCEV *MinIterSCEV = BunldedSCEV.AddRecExpr->evaluateAtIteration(
+      SE.getConstant(APInt(32, MinIterCount)), SE);
+
+  // If the MinIterSCEV is not a constant, but based on an expression,
+  // the no signed/unsigned overflow flags have to be added, so that the
+  // expression can be simplified in later passes
+  SCEVCouldNotCompute::NoWrapFlags NWF = BunldedSCEV.AddRecExpr->getNoWrapFlags(
+      static_cast<SCEVCouldNotCompute::NoWrapFlags>(SCEV::FlagNUW |
+                                                    SCEV::FlagNSW));
+  auto *AE = dyn_cast<SCEVAddExpr>(MinIterSCEV);
+  if (AE && NWF) {
+    MinIterSCEV = SE.getAddExpr(AE->getOperand(0), AE->getOperand(1), NWF);
```

Comment:
Essentially this is taking the same expression and adding flags? Isn't there a helper to add them without re-creating the SCEV? Do you also know why those flags aren't automatically propagated from `BunldedSCEV.AddRecExpr`? After all, `MinIterSCEV` is that same expression, but evaluated at a different point.

Reply:
Yeah, I can add the overflow flags without recreating the expression, I generalized it to add the flags for any 'SCEVCommutativeExpr' 

Unfortunately `SCEVAddRecExpr::evaluateAtIteration` does not have a way to automatically reuse the no overflow flags.